### DISCUSSION TITLE
How to properly exec a long-running database operation during the launching phase?

### DISCUSSION THREAD

**Discussion by tumluliu** [rwf2/Rocket#2187] [Questions] [answered]
Hi all, 

I am trying to make a long-running (~1min) function call which queries the database and constructs some data objects for upcoming usage. But it will give me a runtime panic if I create a `postgres::Client` within the launch function.

```
thread 'main' panicked at 'Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks.'
```

a minimal sample code snippet will be like this:

```
use postgres::{Client, NoTls};
use my_crate::my_libs::fill_data_model_from_db;

#[macro_use] extern crate rocket;

#[get("/world")]
fn world() -> &'static str {
    "Hello, world!"
}

#[launch]
fn rocket() -> _ {
    let mut client: Client = Client::connect("host=localhost dbname=testdb", NoTls).unwrap();
    // long running db queries for data preparation
    // let d = fill_data_model_from_db(&mut client);
    // will use State to encapsulate `d`
    rocket::build().mount("/hello", routes![world])
}
```

It's a bit strange to me that although the `postgres::Client` is actually a `synchronous PostgreSQL client` as its document says, why the creation of it will cause a nested tokio runtime issue? I already took a look at the [rocket_db_pools](https://api.rocket.rs/v0.5-rc/rocket_db_pools/index.html), but could not figure out how the db pool would help here. I'm still pretty new to rust and rocket, and looking forward to your hints. Thanks a lot in advance!

**Comment by the10thWiz** [comment, marked as answer]
> It's a bit strange to me that although the `postgres::Client` is actually a synchronous `PostgreSQL` client as its document says, why the creation of it will cause a nested tokio runtime issue?

Looking into the `postgres` crate, it looks like it starts it's own async runtime and runs an async client on the runtime it creates. It's technically a sync client (since it doesn't need an async runtime), but it can't actually be started within a tokio runtime due to some things tokio does. The simple solution here would actually be to use something like `tokio-postgres`, an async client (since Rocket already creates a tokio runtime, you might as well use it). The `postgres` crate actually uses `tokio-postgres` behind the scenes, so it shouldn't be too hard to translate your code to work with the async runtime.

**Reply by SergioBenitez** [reply]
To add on, instead of using tokio-Postgres, use rocket_db_pools instead.

**Reply by tumluliu** [reply]
Thank you both for your explanations and suggestions!
